### PR TITLE
run dockerfile examples tests

### PR DESCRIPTION
**Pull request by alex1545**

### CODE REVIEW

File: testing/e2e.sh [hunk 1]
```diff
@@ -542,6 +534,36 @@ function test_py_image_deps_as_layers() {
...
+function test_dockerfile_image_apt_pkgs() {
+  cd "${ROOT}/testing/examples/run_instruction_apt_pkgs"
+  clear_docker
+  bazel test --host_force_python=PY2 ...
+}
+
+function test_dockerfile_image_arbitrary() {
+  cd "${ROOT}/testing/examples/run_instruction_arbitrary"
+  clear_docker
+  bazel test --host_force_python=PY2 ...
```
Comment (nlopezgi) [inline testing/e2e.sh]: add .bazelrc files to the projects with this flag

**Comment by nlopezgi** [comment]
I'd rather we setup these new tests as GCB triggers if its not too much additional effort to do so.

File: testing/e2e.sh (hunk 1 above)
Comment (alex1545) [reply testing/e2e.sh]: Done.
There is only one "internal" project for all examples, so added one such file.

**Comment by alex1545** [comment]
These tests will run as GCB jobs. Removed the tests from the `e2e.sh` script and added `cloudbuild.yaml` files for each of the examples. Single `cloudbuild.yaml` file won't be able to run the examples in parallel because it looks like bazel can be launched at most 4 times at a time.

These tests won't run for now, until this repo is configured to use GCB triggers for its PRs.

File: contrib/dockerfile_build.bzl [hunk 2]
```diff
@@ -89,7 +89,7 @@ exports_files(["{}"])
         if build_result.return_code:
             fail("Image build command failed: {} ({})".format(
                 build_result.stderr,
-                " ".join(command),
+                " ".join([str(e) for e in command]),
```
Comment (nlopezgi) [inline contrib/dockerfile_build.bzl]: why is this needed, shouldn't all elements in command already be strings? If not, its probably better to make them strings as you are putting them in the command instead of converting them here

File: testing/examples/WORKSPACE [hunk 3]
```diff
@@ -16,9 +16,11 @@ workspace(name = "dockerfile_examples")
 
 load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive", "http_file")
 
-local_repository(
+http_archive(
```
Comment (nlopezgi) [inline testing/examples/WORKSPACE]: Why the change from local_repository? You should leave it as local repo otherwise you would be testing a committed version of rules_docker not the changes you have in the PR

File: testing/examples/basic/cloudbuild.yaml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Tests that verify the basic example of rules_docker vs Dockerfile.
+
+steps:
+  # Confirm bazel and docker are installed.
+  - name: "gcr.io/asci-toolchain/nosla-ubuntu16_04-bazel-docker-gcloud"
```
Comment (nlopezgi) [inline testing/examples/basic/cloudbuild.yaml]: use the bazel container (or do you need gcloud for some reason)?

File: testing/examples/basic/cloudbuild.yaml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    entrypoint: "sh"
+    args: ["-c", "bazel version && docker version"]
+    id: "versions"
+    waitFor: ["-"]
+
+  # Test the basic example.
+  - name: "gcr.io/asci-toolchain/nosla-ubuntu16_04-bazel-docker-gcloud"
+    entrypoint: "sh"
+    args: ["-c", "cd testing/examples && bazel test //basic:all"]
+    waitFor: ["versions"]
```
Comment (nlopezgi) [inline testing/examples/basic/cloudbuild.yaml]: use 'dir:' to set the directory to testing/examples (Example: https://github.com/bazelbuild/bazel-toolchains/blob/master/tests/rbe_repo/examples/rbe_custom.yaml#L8)
That way you dont need an entrypoint and the only args you should pass are 
 - test 
 - //basic:all

File: testing/examples/basic/cloudbuild.yaml [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Tests that verify the basic example of rules_docker vs Dockerfile.
+
+steps:
+  # Confirm bazel and docker are installed.
+  - name: "gcr.io/asci-toolchain/nosla-ubuntu16_04-bazel-docker-gcloud"
+    entrypoint: "sh"
+    args: ["-c", "bazel version && docker version"]
```
Comment (nlopezgi) [inline testing/examples/basic/cloudbuild.yaml]: No need to have a step just to check the bazel version and docker version

File: testing/examples/run_instruction_apt_pkgs/cloudbuild.yaml [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  - name: "gcr.io/asci-toolchain/nosla-ubuntu16_04-bazel-docker-gcloud"
+    entrypoint: "sh"
+    args: ["-c", "bazel version && docker version"]
+    id: "versions"
+    waitFor: ["-"]
+
+  # Test the run_instruction_apt_pkgs example.
+  - name: "gcr.io/asci-toolchain/nosla-ubuntu16_04-bazel-docker-gcloud"
+    entrypoint: "sh"
+    args: ["-c", "cd ../testing/examples && bazel --output_base=/workspace/apt_pkgs test //run_instruction_apt_pkgs:all"]
```
Comment (nlopezgi) [inline testing/examples/run_instruction_apt_pkgs/cloudbuild.yaml]: Why do you need to set the output base?

File: contrib/dockerfile_build.bzl (hunk 2 above)
Comment (alex1545) [reply contrib/dockerfile_build.bzl]: Done.
Some of the elements were of type `path`.

File: testing/examples/WORKSPACE (hunk 3 above)
Comment (alex1545) [reply testing/examples/WORKSPACE]: Done.
Changed that when had trouble running targets (on GCB) in the "internal" bazel project within the rules_docker project. I figured it out now.

File: testing/examples/basic/cloudbuild.yaml (hunk 4 above)
Comment (alex1545) [reply testing/examples/basic/cloudbuild.yaml]: Done.
No, just need bazel and docker.

File: testing/examples/basic/cloudbuild.yaml (hunk 6 above)
Comment (alex1545) [reply testing/examples/basic/cloudbuild.yaml]: Done.

File: testing/examples/basic/cloudbuild.yaml (hunk 5 above)
Comment (alex1545) [reply testing/examples/basic/cloudbuild.yaml]: Done.

File: testing/examples/run_instruction_apt_pkgs/cloudbuild.yaml (hunk 7 above)
Comment (alex1545) [reply testing/examples/run_instruction_apt_pkgs/cloudbuild.yaml]: Not a 100% sure, but this makes sure that Bazel generated files can be a accessed by multiple containers (even within the same steps). This is needed here and in the run_instruction_arbitrary example because these examples use the `download_pkgs`, `install_pkgs` and `container_run_and_commit` rules that launch additional containers.

Comment (nlopezgi) [reply testing/examples/run_instruction_apt_pkgs/cloudbuild.yaml]: So do you get an error if you dont set this?
What type of error?
If its because these rules need to launch additional containers, there should be some other way to avoid issues; e.g., by passing `/var/run/docker.sock:/var/run/docker.sock` in `volumes` (https://cloud.google.com/cloud-build/docs/build-config)

Comment (alex1545) [reply testing/examples/run_instruction_apt_pkgs/cloudbuild.yaml]: Ass discussed separately, added a comment why this is needed.

File: testing/examples/run_instruction_apt_pkgs/cloudbuild.yaml [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Tests that verify the example installing apt packages using rules_docker
+# and Dockerfile.
+
+timeout: 21600s
```
Comment (nlopezgi) [inline testing/examples/run_instruction_apt_pkgs/cloudbuild.yaml]: Does this really need all that time to run?

File: testing/examples/run_instruction_apt_pkgs/cloudbuild.yaml [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+options:
```
Comment (nlopezgi) [inline testing/examples/run_instruction_apt_pkgs/cloudbuild.yaml]: why does this need the highcpu machine?

File: testing/examples/run_instruction_arbitrary/cloudbuild.yaml [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Tests that verify the example manipulating arbitrary commands using
+# rules_docker and Dockerfile.
+
+timeout: 21600s
```
Comment (nlopezgi) [inline testing/examples/run_instruction_arbitrary/cloudbuild.yaml]: same q as above

File: testing/examples/run_instruction_apt_pkgs/cloudbuild.yaml (hunk 8 above)
Comment (alex1545) [reply testing/examples/run_instruction_apt_pkgs/cloudbuild.yaml]: Done.
Reduced timeout to 30 mins.

File: testing/examples/run_instruction_apt_pkgs/cloudbuild.yaml (hunk 9 above)
Comment (alex1545) [reply testing/examples/run_instruction_apt_pkgs/cloudbuild.yaml]: I thought it was worth it as it bring the time of this test from almost 13 mins to ~5 mins.

File: testing/examples/run_instruction_arbitrary/cloudbuild.yaml (hunk 10 above)
Comment (alex1545) [reply testing/examples/run_instruction_arbitrary/cloudbuild.yaml]: Timeout is reduced.
High CPU here brings the test time from almost 20 mins to just under 10 mins.

File: testing/examples/run_instruction_apt_pkgs/cloudbuild.yaml (hunk 9 above)
Comment (nlopezgi) [reply testing/examples/run_instruction_apt_pkgs/cloudbuild.yaml]: cool, leave with the highcpu then but remove the timeout then!

Comment (alex1545) [reply testing/examples/run_instruction_apt_pkgs/cloudbuild.yaml]: Done.
